include/graphblas/bsp1d/blas3.hpp

```diff
@@ -195,16 +195,118 @@ namespace grb {
 			phase
 		);
 		if( phase == RESIZE ) {
-			if( collectives<>::allreduce( ret, operators::any_or< RC >() ) != SUCCESS ) {
+			if( collectives<>::allreduce(
+					ret, operators::any_or< RC >()
+				) != SUCCESS
+			) {
 				return PANIC;
-			} else {
-				return SUCCESS;
 			}
+			return SUCCESS;
 		}
 		assert( phase == EXECUTE );
 		return internal::checkGlobalErrorStateOrClear( C, ret );
 	}
 
+	template<
+		Descriptor descr = descriptors::no_operation,
+		typename InputType, typename OutputType,
+		typename RIT_L, typename CIT_L, typename NIT_L,
+		typename RIT_A, typename CIT_A, typename NIT_A
+	>
+	RC tril(
+		Matrix< OutputType, BSP1D, RIT_L, CIT_L, NIT_L > &L,
+		const Matrix< InputType, BSP1D, RIT_A, CIT_A, NIT_A > &A,
+		const long int k,
+		const Phase &phase = Phase::EXECUTE,
+		const typename std::enable_if<
+			!grb::is_object< OutputType >::value &&
+			!grb::is_object< InputType >::value &&
+			std::is_convertible< InputType, OutputType >::value
+		>::type * const = nullptr
+	) {
+#ifdef _DEBUG
+		std::cout << "In grb::tril( BSP1D )\n";
+#endif
+		assert( phase != TRY );
+		const RC ret = tril< descr >(
+			internal::getLocal( L ),
+			internal::getLocal( A ),
+			k,
+			phase
+		);
```

Comment:
Simply delegating has an issue that local coordinates may first need to be mapped to global coordinates before we may determine what is above or below the kth diagonal. Here we are using a row-wise block-cyclic distribution, which means local rows have a hidden offset-- e.g., the local row 64 at the first user process actually corresponds to row 128 globally, and therefore it should filter column indices greater-than-or-equal to 128, not 64.

Some primitives like the reference `vxm` for this reason take a hidden `std::function` parameter that allows the reference backend to execute the necessary conversions. An alternative could be a hidden template argument that passes the appropriate `Distribution` class.

A different implementation may instead rely on iterators -- we could iterate over the nonzeroes of `A` in `PARALLEL` I/O mode, meaning we iterate only over the contents of `A` that are locally stored. Iterators already convert local indices to global indices. Thus we can filter the resulting iterator (`operator++` can skip elements that are above the `k`th diagonal), and pass the filtered iterator to a `buildMatrix` routine on `L`. With this type of approach, there are two remarks:

1. `tril` and `triu` could reside in `grb::algorithms`, which frees up new backends from having to implement `tril` and `triu`explicitly (they just have to implement parallel output iteration and matrix ingestion from iterators, which they would have to anyway);

2. however, currently, output matrix iterators are forward iterators, not random access -- we would have to fix that first.

Reply:
I will try to come up with something for the local coordinates issue.

I am not really prone to the idea of transforming `tril+u` into an algorithm.
This would first of all hamper any backend-specific optimisations. 
Moreover, this would not make sense if we plan to generalise `tril+u` into a `filter` primitive using special (boolean) operators capable of handling vector/matrix coordinates (like in other GraphBLAS implementations)